Sources/KSCrashRecordingCore/KSBinaryImageCache.c

```diff
@@ -137,71 +62,55 @@ void ksbic_init(void)
 
     KSLOG_DEBUG("Initializing binary image cache");
 
-    _dyld_register_func_for_add_image(ksbic_addImageCallback);
-    _dyld_register_func_for_remove_image(ksbic_removeImageCallback);
-}
-
-// For testing purposes only. Used with extern in test files.
-void ksbic_resetCache(void)
-{
-    bool expected = true;
-    if (!atomic_compare_exchange_strong(&g_initialized, &expected, false)) {
+    struct task_dyld_info dyld_info;
+    mach_msg_type_number_t count = TASK_DYLD_INFO_COUNT;
+    kern_return_t err = task_info(mach_task_self(), TASK_DYLD_INFO, (task_info_t)&dyld_info, &count);
```

Comment:
As far as I understand, this solution assumes that task_info is a lock-free function. However, after reviewing XNU’s code, I don't believe this is the case:
https://github.com/apple-oss-distributions/xnu/blob/e3723e1f17661b24996789d8afc084c0c3303b26/osfmk/kern/task.c#L5312-L5327

Reply:
Update: Apologies, I took another look at the code and now see that this is actually a solid solution. The scenario I was worried about could only occur during KSCrash installation, so it shouldn’t be a concern here.